### PR TITLE
Facilitate using network node pressures from restart

### PR DESCRIPTION
**Pull request by vkip**

### CODE REVIEW

**Comment by vkip** [comment]
To be used with https://github.com/OPM/opm-simulators/pull/4961. Opening for review to start the discussion on whether EclipseState is the right place for this..

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4961 please

**Comment by GitPaean** [comment]
This PR and the downstream PR https://github.com/OPM/opm-simulators/pull/4961 look fine to me.  To be on the safe side, I am inviting @bska for a second look, since it is on the RESTART facility.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4961 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4961 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4961 please

**Comment by GitPaean** [comment]
Not totally sure what the following jenkins error means, while it might mean the cases for the regression test might not be able to restarted properly even without the network options. 

```
2 keywords exhibit failures
	DOUBHEAD: Restart, sequence 7
		Fails for 1 entries
		Largest absolute error: 9.9999000e+02
		Largest relative error: 9.9999000e-01
	THRESHPR: Restart, sequence 7
		Fails for 2 entries
		Largest absolute error: 2.4868826e+00
		Largest relative error: -1.0000000e+00
```

And other summary comparison also needs to be checked out, especially some values are missing causing 100% relative difference.
```
12 keywords exhibit failures
	WGPGR:PROD1: Summary file
		Fails for 1 entries
		Largest absolute error: 3.1840488e+05
		Largest relative error: -1.0000000e+00
.....
	WGPGR:PROD3: Summary file
		Fails for 1 entries
		Largest absolute error: 4.4533341e+05
		Largest relative error: -1.0000000e+00
.......
	WOPGR:PROD3: Summary file
		Fails for 1 entries
		Largest absolute error: 4.0806199e+03
		Largest relative error: -1.0000000e+00
.....
```


I tried to test the RESTART setup from this PR, but locally I could not run the restart due to the segmentation fault in the following code, **result[idx] = thpres.getThresholdPressure(i, j);** in the following function 

```c++
template<class Grid, class GridView, class ElementMapper, class Scalar>
std::vector<Scalar>
EclGenericThresholdPressure<Grid,GridView,ElementMapper,Scalar>::
getRestartVector() const
{
    if (!enableThresholdPressure_)
        return {};

    std::vector<Scalar> result(numEquilRegions_ * numEquilRegions_, 0.0);
    const auto& simConfig = eclState_.getSimulationConfig();
    const auto& thpres = simConfig.getThresholdPressure();

    std::size_t idx = 0;
    for (unsigned j = 1; j <= numEquilRegions_; ++j) {
        for (unsigned i = 1; i <= numEquilRegions_; ++i, ++idx) {
            if (thpres.hasRegionBarrier(i, j)) {
                if (thpres.hasThresholdPressure(i, j)) {
                    result[idx] = thpres.getThresholdPressure(i, j);
                } else {
                    result[idx] = this->thpresDefault_[idx];
                }
            }
        }
    }

    return result;
}
```


Can you double check how this PR runs on your original setup? And also @bska might understand these issues better and can be helpful. @vkip

**Comment by vkip** [comment]
This appears to be related to restart with threshold pressures, so nothing to do with this PR, except that the test cases were unfortunate enough to use them. The test cases run locally, but with differences in the .UNRST files as shown on Jenkins (I only checked the network relevant parts previously..). I can reproduce the segfault if THPRES is not commented out in the restart deck.

Hopefully @bska can have a look at this, alternatively I guess we can update the test cases to not use THPRES..

**Comment by GitPaean** [comment]
> alternatively I guess we can update the test cases to not use THPRES

If THPRES is not essential to the cases, I think it will be good to remove them so that the RESTART focus on the network related. 

THPRES related RESTART can be handled as a separate issue.

**Comment by vkip** [comment]
> > alternatively I guess we can update the test cases to not use THPRES
> 
> If THPRES is not essential to the cases, I think it will be good to remove them so that the RESTART focus on the network related.
> 
> THPRES related RESTART can be handled as a separate issue.

Ok, appears to be something else here as well - also get issues with current master, even without THPRES and network. Will dig a bit..

**Comment by bska** [comment]
> Hopefully @bska can have a look at this, alternatively I guess we can update the test cases to not use THPRES..

I haven't looked at the details here&ndash;I will do so shortly&ndash;but I'll just comment on the threshold pressure issue.  I noticed this problem whilst working on another restart issue which eventually became #3731.  I **believe** the underlying issue is that we do read the pertinent values from the restart file, but fail to properly form the full internal simulator representation of the threshold pressures from those values.  If at all possible, I'd try to avoid threshold pressure complications for restart here.

**Comment by vkip** [comment]
jenkins build this opm-simulators=4961 please

File: src/opm/io/eclipse/rst/network.cpp [hunk 1]
```diff
@@ -193,6 +193,7 @@ namespace {
                 usys.to_si(Opm::UnitSystem::measure::pressure,
                            rnode[VI::RNode::index::PressureLimit]);
         }
+        node.pressure = rnode[VI::RNode::index::NodePres];
```
Comment (bska) [inline src/opm/io/eclipse/rst/network.cpp:196]: Might you need unit conversion here?  The `rnode` values are **typically** in output units so depending on what code uses the `Node::pressure` value there could be an inconsistency here.

**Review by bska — COMMENTED** [review]
This looks good for the most part, but I **think** we may be missing a unit conversion for the node pressures that are loaded from the restart file.  Unless the client code really expects pressure values in output units, there should be a call to `UnitSystem::to_si(pressure, x)` in the code that internalises those values.

File: src/opm/io/eclipse/rst/network.cpp (hunk 1 above)
Comment (vkip) [reply src/opm/io/eclipse/rst/network.cpp:196]: I believe you are right, will fix. Thanks!

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4961 opm-tests=1092 please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me and I'll merge into master.

**Pull request merged**